pkg/operator/operator.go

```diff
@@ -0,0 +1,191 @@
+package operator
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/cluster-dns-operator/pkg/manifests"
+	operatorclient "github.com/openshift/cluster-dns-operator/pkg/operator/client"
+	operatorconfig "github.com/openshift/cluster-dns-operator/pkg/operator/config"
+	operatorcontroller "github.com/openshift/cluster-dns-operator/pkg/operator/controller"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	kconfig "sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// Operator is the scaffolding for the dns operator. It sets up dependencies
+// and defines the topology of the operator and its managed components, wiring
+// them together.
+type Operator struct {
+	manifestFactory *manifests.Factory
+
+	manager manager.Manager
+	caches  []cache.Cache
+	client  client.Client
+}
+
+// New creates (but does not start) a new operator from configuration.
+func New(config operatorconfig.Config) (*Operator, error) {
+	kubeConfig, err := kconfig.GetConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get kube config: %v", err)
+	}
+	scheme := operatorclient.GetScheme()
+	operatorManager, err := manager.New(kubeConfig, manager.Options{
+		Scheme: scheme,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create operator manager: %v", err)
+	}
+
+	// Create and register the operator controller with the operator manager.
+	operatorController, err := operatorcontroller.New(operatorManager, operatorcontroller.Config{
+		ManifestFactory:        &manifests.Factory{},
+		KubeConfig:             kubeConfig,
+		CoreDNSImage:           config.CoreDNSImage,
+		OpenshiftCLIImage:      config.OpenshiftCLIImage,
+		OperatorReleaseVersion: config.OperatorReleaseVersion,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create operator controller: %v", err)
+	}
+
+	// Create additional controller event sources from informers in the managed
+	// namespace. Any new managed resources outside the operator's namespace
+	// should be added here.
+	mapper, err := apiutil.NewDiscoveryRESTMapper(kubeConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get API Group-Resources")
+	}
+	operandCache, err := cache.New(kubeConfig, cache.Options{Namespace: "openshift-dns", Scheme: scheme, Mapper: mapper})
```

Comment:
I think we could configure the manager to watch the `openshift-dns` namespace instead of creating a separate cache. 

I am guessing you are creating `operandCache` because cluster-ingress-operator does.  In the case of cluster-ingress-operator, the operator watches ingresscontrollers in `openshift-ingress-operator` and deployments and services in the `openshift-ingress` namespace, so because it is watching namespace-scoped resources in two different namespaces, it needs two different caches.  However, because dnses are cluster-scoped, cluster-dns-operator only needs to watch one namespace (`openshift-dns`), so I think we can do without `operandCache` in cluster-dns-operator if we configure the manager for `openshift-dns`.

Reply:
We can do this in a follow-up though.

Reply:
I will submit a follow-up pr for this.

Reply:
Covered by https://github.com/openshift/cluster-dns-operator/pull/87